sklearn/cluster/dbscan_.py

```diff
@@ -118,18 +125,30 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
     if metric == 'precomputed':
         D = pairwise_distances(X, metric=metric)
         neighborhoods = np.empty(X.shape[0], dtype=object)
-        neighborhoods[:] = [np.where(x <= eps)[0] for x in D]
+        if sparse.issparse(D):
+            D_mask = D.data <= eps
+            masked_indices = D.indices.astype(np.intp)[D_mask]
+            masked_indptr = np.cumsum(D_mask)[D.indptr[1:] - 1]
+            # insert the diagonal
+            masked_indices = np.insert(masked_indices, masked_indptr,
+                                       np.arange(D.shape[0]))
+            masked_indptr = masked_indptr[:-1] + np.arange(1, D.shape[0])
+            # split into rows
+            neighborhoods[:] = np.split(masked_indices, masked_indptr)
+        else:
+            neighborhoods[:] = [np.where(x <= eps)[0] for i, x in enumerate(D)]
```

Comment:
Relic of an alternative implementation that dropped the diagonal out of these neighbors instead and used `radius_neighbors_graph(X=None, ...)` below. I decided to be less disruptive.